### PR TITLE
Add preload paths for autosaves

### PR DESCRIPTION
**Pull request by talldan**
## Description
Adds a preload path for autosaves to help optimize the changes made in #7945.

Note there is a trac ticket that replicates this functionality in core (https://core.trac.wordpress.org/ticket/46974). In the meantime this should bring the same optimization to the plugin.

## How has this been tested?
1. Create a new post and publish it
2. Open the browser dev tool and monitor the network requests
2. Reload the editor

**With this change:** no requests to the autosaves endpoint are made when reloading the editor.
**Without this change:** a request is made to the autosaves endpoint when reloading the editor.

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Review by aduth — CHANGES_REQUESTED** [review]
Prior to this I had opened #15061 for a similar need. It's since been merged. There are now conflicts here, though I think it should be simple enough to follow the same pattern.

**Comment by talldan** [comment]
@aduth - now rebased. Merging the changes wasn't too bad, the tests were a bit trickier, but I've managed to work it out.

File: phpunit/class-extend-preload-paths-test.php [hunk 1]
```diff
@@ -7,20 +7,48 @@
...
+	public static function wpSetUpBeforeClass( $factory ) {
+		self::$post = $factory->post->create_and_get();
+	}
+
+	/**
+	 * Tests paths added if missing.
 	 */
 	function test_localizes_script() {
-		$preload_paths = gutenberg_extend_block_editor_preload_paths( array() );
+		$post = (object) array(
```
Comment (aduth) [inline phpunit/class-extend-preload-paths-test.php]: Is this variable (and in the test below) unused?

File: lib/client-assets.php [hunk 2]
```diff
@@ -591,13 +591,30 @@ function gutenberg_extend_block_editor_styles( $settings ) {
...
+	 *
+	 * Trac ticket: https://core.trac.wordpress.org/ticket/46974
+	 *
+	 * At the time of writing, the change is not committed or released
+	 * in core. This path should be removed from Gutenberg when the code is
+	 * released in core, and the corresponding release version becomes
+	 * the minimum supported version.
+	 */
+	$post_type_object = get_post_type_object( $post->post_type );
+	$rest_base        = ! empty( $post_type_object->rest_base ) ? $post_type_object->rest_base : $post_type_object->name;
```
Comment (aduth) [inline lib/client-assets.php]: While it seems unlikely given the filter flow, it makes me nervous to see `null` as a possible return type of `get_post_type_object` and then to have unguarded object property references. Should we be more guarded (wrap this whole bit in an `if ( ! is_null( $post_type_object ) ) {`)?

File: phpunit/class-extend-preload-paths-test.php (hunk 1 above)
Comment (talldan) [reply phpunit/class-extend-preload-paths-test.php]: Yep, I made a mistake leaving that there after discovering that I could use the post factory in the test. Thanks for pointing it out.

File: lib/client-assets.php (hunk 2 above)
Comment (talldan) [reply lib/client-assets.php]: 👍 good idea, I've addressed that.

**Review by aduth — APPROVED** [review]
👍

**Comment by youknowriad** [comment]
It looks like we need to commit this patch as well https://core.trac.wordpress.org/ticket/46974 for WP 5.3

**Pull request merged**